doc/developer/design/20231027_refresh_every_mv.md

```diff
@@ -0,0 +1,154 @@
+
+### Implementation Alternative: Creating a New Dataflow at Each Refresh
+
+(The above alternatives are alternatives to `REFRESH EVERY` MVs, but this alternative is just about the implementation details of `REFRESH EVERY` MVs.)
+
+This approach would need similar replica management as the proposed solution, but instead of rehydrating an existing dataflow, we would always create a new dataflow to run the refresh. An advantage of this approach is that these dataflows could be single-time dataflows, and thus rely on one-shot SELECT optimizations. However, this approach would be more complicated, as we would have to separate the lifecycle of dataflows from the lifecycle of MVs. These new dataflows would be a kind of hybrid between Peek dataflows and MV dataflows, and would require new code in various parts of the Compute Controller. As we are under some time pressure to implement this feature, we opted for the simpler solution proposed in this document.
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Open questions
+
+### Special Clusters for `REFRESH EVERY` MVs
+
+I'm still thinking about how exactly to tie a `REFRESH EVERY` MV to the cluster whose replicas will be managed by the MV. I can see two approaches:
+- These MVs would automatically manage not just replicas, but even their cluster. That is, the system would create a cluster when creating the MV, and drop the cluster when dropping the MV. (And this cluster could not be used for anything else.) The problem with this approach is that then certain cluster creation options (e.g, size) would have to be copied from `CREATE CLUSTER` to `CREATE MATERIALIZED VIEW`. I'm not sure how big of an issue this would be.
+- Another approach would be to create `REFRESH EVERY` MVs simply on the active cluster (as other, normal MVs), and the user would be expected to separately create a cluster manually. A further tweak could be to let the user specify a special option at cluster creation, which would make the system let only `REFRESH EVERY` MVs be created on this cluster (possibly only with the exact same refresh settings). This would avoid situations where a user accidentally creates some other, normal object on a cluster whose replicas are being managed by a `REFRESH EVERY` MV, and then this other object would unexpectedly also stop refreshing.
```

Comment:
Alternatively: Install `REFRESH` MVs on existing clusters like regular MVs, but automatically spin down cluster replicas only when the cluster has only idle `REFRESH` MVs. This also ties in well to a possible future auto-scaling feature where the controller would check if a replica has any work to do and spin it down otherwise.

To prevent users from accidentally keeping their expensive refresh replica running, we could produce a notice when one create a refresh MV on a cluster with other dataflows, or vice versa. Or simply forbid the creation of both of these things on the same cluster, as we do today for compute and storage objects.

Reply:
Added to the text.